os/bluestore: fix zombie spanning blobs.

ifed01 opened the pull request:
Fixes: https://tracker.ceph.com/issues/69646

See 
https://tracker.ceph.com/issues/69646?next_issue_id=69645&prev_issue_id=69647#note-6
for RCA.




<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

ifed01 committed:
os/bluestore: fix zombie spanning blobs.

Fixes: https://tracker.ceph.com/issues/69646

Signed-off-by: Igor Fedotov <igor.fedotov@croit.io>

ifed01 commented:
jenkins test api

ifed01 commented:
jenkins render docs

ifed01 commented:
jenkins docs test

ifed01 commented:
jenkins test docs

aclamk commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -3814,11 +3823,18 @@ void BlueStore::ExtentMap::reshard(
 		  dout(20) << __func__ << "    splitting blob, bstart 0x"
```
Lets observe that function `blob_escapes_range` does not work as I would expect.
It is not looking if outside range is anything allocated.
As for high end (right side I call it) Blob should be truncated and there should be no mapping, so its OK.
The low end (left side I call it) can have unmapped regions.
I think that some of the problems with split_blob could be avoided if we always split blob that produces non-degenerate results.

aclamk commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -4502,29 +4518,103 @@ BlueStore::BlobRef BlueStore::ExtentMap::split_blob(
   BlobRef rb = onode->c->new_blob();
```
I think we should only split Blob along offset that produces 2 non-empty Blobs.
After that walk through extents modifies some of them to use newly created Blob.

The check if Blob are OK should be a separate function.
It can be applied before of after split.
But it feels like extra sanity check, not a part of split_blob logic.

If we make live (non-fsck) check and repair functions, they should be a standalone entities, so we could call them from different contexts. It would be an interesting addition for debug / research, to verify metadata is valid.

ifed01 replied:
I think I've updated the code in a way you asked. ExtentMap::maybe_normalize() method is a standalone checker/repairer for  dangling extents.

ifed01 committed:
os/bluestore: simplify BlueStore::reshard_action() method

This patch separates dump logic for onode with too many spanning blobs
into a standalone function.

Signed-off-by: Igor Fedotov <igor.fedotov@croit.io>

ifed01 committed:
os/bluestore: split dangling extents detection off split_blob() func.

Signed-off-by: Igor Fedotov <igor.fedotov@croit.io>

ifed01 commented:
@aclamk - I think this is ready for another review round.

ifed01 commented:
This has been (hopefully) fix by https://github.com/ceph/ceph/pull/68798
